crates/red_knot_module_resolver/src/path.rs

```diff
@@ -0,0 +1,693 @@
+                let file_system = db.file_system();
+                file_system.exists(&fs_path.join("__init__.py"))
+                    || file_system.exists(&fs_path.join("__init__.pyi"))
+            }
+            // Unlike the other variants:
+            // (1) Account for VERSIONS
+            // (2) Only test for `__init__.pyi`, not `__init__.py`
+            (Self::StandardLibrary(StandardLibraryPath(fs_path)), Self::StandardLibrary(stdlib_root)) => {
+                let Some(module_name) = self.as_module_name() else {
+                    return false;
+                };
+                let typeshed_versions = Self::load_typeshed_versions(db, stdlib_root);
+                match typeshed_versions.query_module(&module_name, get_target_py_version(db)) {
+                    TypeshedVersionsQueryResult::Exists
+                    | TypeshedVersionsQueryResult::MaybeExists => {
+                        db.file_system().exists(&fs_path.join("__init__.pyi"))
+                    }
+                    TypeshedVersionsQueryResult::DoesNotExist => false,
+                }
+            }
+            (path, root) => unreachable!(
+                "The search path should always be the same variant as `self` (got: {path:?}, {root:?})"
+            )
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn parent(&self) -> Option<Self> {
+        Some(match self {
+            Self::Extra(ExtraPath(path)) => Self::extra_unchecked(path.parent()?),
+            Self::FirstParty(FirstPartyPath(path)) => Self::first_party_unchecked(path.parent()?),
+            Self::StandardLibrary(StandardLibraryPath(path)) => {
+                Self::standard_library_unchecked(path.parent()?)
+            }
+            Self::SitePackages(SitePackagesPath(path)) => {
+                Self::site_packages_unchecked(path.parent()?)
+            }
+        })
+    }
+
+    #[must_use]
+    fn ends_with_dunder_init(&self) -> bool {
+        match self {
+            Self::Extra(ExtraPath(path))
+            | Self::FirstParty(FirstPartyPath(path))
+            | Self::SitePackages(SitePackagesPath(path)) => {
+                path.ends_with("__init__.py") || path.ends_with("__init__.pyi")
+            }
+            Self::StandardLibrary(StandardLibraryPath(path)) => path.ends_with("__init__.pyi"),
+        }
+    }
+
+    #[must_use]
+    fn sans_dunder_init(self) -> Self {
+        if self.ends_with_dunder_init() {
+            self.parent().unwrap_or_else(|| match self {
+                Self::Extra(_) => Self::extra_unchecked(""),
+                Self::FirstParty(_) => Self::first_party_unchecked(""),
+                Self::StandardLibrary(_) => Self::standard_library_unchecked(""),
+                Self::SitePackages(_) => Self::site_packages_unchecked(""),
+            })
+        } else {
+            self
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn as_module_name(&self) -> Option<ModuleName> {
```

Comment:
The Rust convention is to call this `to_module_name` because the conversation isn't "free". See [naming guidelines](https://rust-lang.github.io/api-guidelines/naming.html#ad-hoc-conversions-follow-as_-to_-into_-conventions-c-conv)

Reply:
I tried making this change, but Clippy complained:

```
warning: methods with the following characteristics: (`to_*` and `self` type is `Copy`) usually take `self` by value
   --> crates/red_knot_module_resolver/src/path.rs:406:34
    |
406 |     pub(crate) fn to_module_name(&self) -> Option<ModuleName> {
    |                                  ^^^^^
    |
    = help: consider choosing a less ambiguous name
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#wrong_self_convention
    = note: `#[warn(clippy::wrong_self_convention)]` on by default
```

Reply:
This is because your type is copy, so you have to change from `&self` to `self`